File: frontend/src/api/tactico.ts
import { api } from "@/lib/api";
import type { TacticoMatch, TacticoSearchStatus, TacticoSquad, TacticoStats } from "@/types";

export async function fetchTacticoStats(): Promise<TacticoStats> {
  const { data } = await api.get<TacticoStats>("/tactico/stats");
  return data;
}

export async function fetchTacticoSquad(): Promise<TacticoSquad> {
  const { data } = await api.get<TacticoSquad>("/tactico/squad");
  return data;
}

export async function setTacticoSquad(userCardIds: number[]): Promise<TacticoSquad> {
  const { data } = await api.put<TacticoSquad>("/tactico/squad", { user_card_ids: userCardIds });
  return data;
}

export async function createTacticoBotMatch(difficulty: "easy" | "medium" | "hard"): Promise<TacticoMatch> {
  const { data } = await api.post<TacticoMatch>("/tactico/matches/bot", { difficulty });
  return data;
}

export async function createTacticoChallenge(receiverId: number): Promise<TacticoMatch> {
  const { data } = await api.post<TacticoMatch>("/tactico/matches/challenge", { receiver_id: receiverId });
  return data;
}

export async function acceptTacticoChallenge(id: number): Promise<TacticoMatch> {
  const { data } = await api.post<TacticoMatch>(`/tactico/matches/${id}/accept`);
  return data;
}

export async function declineTacticoChallenge(id: number): Promise<TacticoMatch> {
  const { data } = await api.post<TacticoMatch>(`/tactico/matches/${id}/decline`);
  return data;
}

export async function cancelTacticoChallenge(id: number): Promise<TacticoMatch> {
  const { data } = await api.post<TacticoMatch>(`/tactico/matches/${id}/cancel`);
  return data;
}

export async function submitTacticoRound(id: number, userCardId: number): Promise<TacticoMatch> {
  const { data } = await api.post<TacticoMatch>(`/tactico/matches/${id}/rounds`, { user_card_id: userCardId });
  return data;
}

export async function forfeitTacticoMatch(id: number): Promise<TacticoMatch> {
  const { data } = await api.post<TacticoMatch>(`/tactico/matches/${id}/forfeit`);
  return data;
}

export async function fetchTacticoMatches(): Promise<TacticoMatch[]> {
  const { data } = await api.get<TacticoMatch[]>("/tactico/matches");
  return data;
}

export async function fetchTacticoMatch(id: number): Promise<TacticoMatch> {
  const { data } = await api.get<TacticoMatch>(`/tactico/matches/${id}`);
  return data;
}

export async function startTacticoSearch(): Promise<TacticoSearchStatus> {
  const { data } = await api.post<TacticoSearchStatus>("/tactico/matchmaking/search");
  return data;
}

export async function fetchTacticoSearchStatus(): Promise<TacticoSearchStatus> {
  const { data } = await api.get<TacticoSearchStatus>("/tactico/matchmaking/status");
  return data;
}

export async function cancelTacticoSearch(): Promise<void> {
  await api.post("/tactico/matchmaking/cancel");
}

File: backend/app/routers/tactico.py
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_current_user
from app.core.rate_limit import check_rate_limit
from app.database import get_db
from app.models.user import User
from app.schemas.tactico import (
    TacticoBotMatchRequest,
    TacticoChallengeRequest,
    TacticoMatchOut,
    TacticoRoundSubmitRequest,
    TacticoSearchStatusOut,
    TacticoSquadOut,
    TacticoSquadSetRequest,
    TacticoStatsOut,
)
from app.services import tactico_service

router = APIRouter(prefix="/tactico", tags=["tactico"])


@router.get("/stats", response_model=TacticoStatsOut)
async def get_stats(db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await tactico_service.get_stats(db, user)


@router.get("/squad", response_model=TacticoSquadOut)
async def get_squad(db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await tactico_service.get_squad(db, user)


@router.put("/squad", response_model=TacticoSquadOut)
async def set_squad(
    payload: TacticoSquadSetRequest, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)
):
    return await tactico_service.set_squad(db, user, payload.user_card_ids)


@router.post("/matches/bot", response_model=TacticoMatchOut)
async def create_bot_match(
    payload: TacticoBotMatchRequest, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)
):
    check_rate_limit(f"tactico_bot_match:{user.id}", max_calls=15, window_seconds=60)
    return await tactico_service.create_bot_match(db, user, payload.difficulty)


@router.post("/matches/challenge", response_model=TacticoMatchOut)
async def create_challenge(
    payload: TacticoChallengeRequest, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)
):
    check_rate_limit(f"tactico_challenge:{user.id}", max_calls=10, window_seconds=60)
    return await tactico_service.create_challenge(db, user, payload.receiver_id)


@router.post("/matches/{match_id}/accept", response_model=TacticoMatchOut)
async def accept_challenge(match_id: int, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await tactico_service.accept_challenge(db, user, match_id)


@router.post("/matches/{match_id}/decline", response_model=TacticoMatchOut)
async def decline_challenge(match_id: int, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await tactico_service.decline_challenge(db, user, match_id)


@router.post("/matches/{match_id}/cancel", response_model=TacticoMatchOut)
async def cancel_challenge(match_id: int, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await tactico_service.cancel_challenge(db, user, match_id)


@router.post("/matches/{match_id}/forfeit", response_model=TacticoMatchOut)
async def forfeit_match(match_id: int, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await tactico_service.forfeit_match(db, user, match_id)


@router.post("/matches/{match_id}/rounds", response_model=TacticoMatchOut)
async def submit_round(
    match_id: int,
    payload: TacticoRoundSubmitRequest,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    return await tactico_service.submit_round(db, user, match_id, payload.user_card_id)


@router.get("/matches", response_model=list[TacticoMatchOut])
async def list_matches(db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await tactico_service.list_matches(db, user)


@router.get("/matches/{match_id}", response_model=TacticoMatchOut)
async def get_match(match_id: int, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await tactico_service.get_match(db, user, match_id)


@router.post("/matchmaking/search", response_model=TacticoSearchStatusOut)
async def start_search(db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    check_rate_limit(f"tactico_search:{user.id}", max_calls=10, window_seconds=60)
    entry = await tactico_service.start_search(db, user)
    return TacticoSearchStatusOut(status="searching", match_id=None, created_at=entry.created_at)


@router.get("/matchmaking/status", response_model=TacticoSearchStatusOut)
async def get_search_status(db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    status, match_id = await tactico_service.get_search_status(db, user)
    return TacticoSearchStatusOut(status=status, match_id=match_id, created_at=None)


@router.post("/matchmaking/cancel", status_code=204)
async def cancel_search(db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    await tactico_service.cancel_search(db, user)
